compiler/passes/insertLineNumbers.cpp

```diff
@@ -77,356 +78,393 @@ int getFilenameLookupPosition(std::string name) {
+    VarSymbol* fileLocal = newTemp("_fn", dtInt[INT_SIZE_32]);
 
-    } else {
-      // Apply the line number from the call AST node
-      line = new_IntSymbol(call->linenum());
+    fn->insertAtHead("'move'(%S, '.v'(%S, %S))", fileLocal, bundleArg->sym,
+                     fileField);
+    fn->insertAtHead(new DefExpr(fileLocal));
 
-      int filenameIdx = getFilenameLookupPosition(call->fname());
+    result = {lineLocal, fileLocal};
 
-      file = new_IntSymbol(filenameIdx, INT_SIZE_32);
-    }
   } else {
-    // This branch handles the case in which we should be using
-    // file/line arguments in the enclosing function.
+    ArgSymbol* line =
+        new ArgSymbol(INTENT_CONST_IN, "_ln", dtInt[INT_SIZE_DEFAULT]);
+    fn->insertFormalAtTail(line);
 
-    // Create them if they don't exist yet.
-    if (lineArg == NULL)
-      lineArg = newLine(fn);
-    if (fileArg == NULL)
-      fileArg = newFile(fn);
+    ArgSymbol* file = new ArgSymbol(INTENT_IN, "_fn", dtInt[INT_SIZE_32]);
+    fn->insertFormalAtTail(file);
 
-    line = lineArg;
-    file = fileArg;
+    result = {line, file};
   }
 
-  if (call->isPrimitive(PRIM_GET_USER_FILE) ||
-      call->isPrimitive(PRIM_GET_USER_LINE)) {
+  lineAndFilenameMap.emplace_hint(it, fn, result);
 
-    // Replace these primitives with whatever line/file we chose
-    // (probably an argument)
-    if (call->isPrimitive(PRIM_GET_USER_FILE)) {
-      call->replace(new SymExpr(file));
-    } else if (call->isPrimitive(PRIM_GET_USER_LINE)) {
-      call->replace(new SymExpr(line));
-    }
+  for (CallExpr* call : *fn->calledBy) {
+    enqueue(call->getFunction(), call);
+  }
 
-  } else {
-    // add line/file as arguments to the call
-    call->insertAtTail(line);
-    call->insertAtTail(file);
+  if (auto* others = rootsOrChildren(fn)) {
+    for (FnSymbol* fn : *others) {
+      fn->addFlag(FLAG_INSERT_LINE_FILE_INFO);
+      enqueue(fn);
+    }
   }
+
+  return result;
 }
 
+//
+// InsertLineNumbers works on a function and has to take care of
+//   * modifying our own function (if necessary) to take line/file formals
+//   * modifying calls in our body that need line/file actuals
+//
+void InsertLineNumbers::process(FnSymbol *fn) {
+  // Refactor note: this precondition is a general check that could be hoisted
+  precondition(fn);
+
+  // if we must have line/file info formals
+  if (fn->hasFlag(FLAG_INSERT_LINE_FILE_INFO) &&
+      !fn->hasFlag(FLAG_LINE_NUMBER_OK)) {
+    // Explicitly ignore the return result since we don't need it right now. Our
+    // call sites will be enqueued
+    std::ignore = getLineAndFileForFn(fn);
+  }
 
-static bool isClassMethodCall(CallExpr* call) {
-  FnSymbol* fn     = call->resolvedFunction();
-  bool      retval = false;
+  // We only need to process each call in our body once; calls which may later
+  // require line numbers will be enqueued by their function.
+  // We may need to process the fn itself more than once, so that if we discover
```

Comment:
I think it would have helped me snap to what this comment was saying faster if it said something like "This check is after the above getLineAndFileForFn call so that..."

Reply:
updated